src/views/SafeAreaView.js

```diff
@@ -0,0 +1,225 @@
+import {
+  DeviceInfo,
+  Dimensions,
+  InteractionManager,
+  NativeModules,
+  Platform,
+  SafeAreaView,
+  View,
+} from 'react-native';
+import withOrientation from './withOrientation';
+
+const { isIPhoneX_deprecated } = DeviceInfo;
+// See https://mydevice.io/devices/ for device dimensions
+const X_WIDTH = 375;
+const X_HEIGHT = 812;
+const PAD_WIDTH = 768;
+const PAD_HEIGHT = 1024;
+
+const { height: D_HEIGHT, width: D_WIDTH } = Dimensions.get('window');
+
+const { PlatformConstants = {} } = NativeModules;
+const { minor = 0 } = PlatformConstants.reactNativeVersion || {};
+
+const isIPhoneX = (() => {
+  if (minor >= 50) {
+    return isIPhoneX_deprecated;
+  }
+
+  return (
+    Platform.OS === 'ios' &&
+    ((D_HEIGHT === X_HEIGHT && D_WIDTH === X_WIDTH) ||
+      (D_HEIGHT === X_WIDTH && D_WIDTH === X_HEIGHT))
+  );
+})();
+
+const isIPad = (() => {
+  if (Platform.OS !== 'ios' || isIPhoneX) return false;
+
+  // if portrait and width is smaller than iPad width
+  if (D_HEIGHT > D_WIDTH && D_WIDTH < PAD_WIDTH) {
+    return false;
+  }
+
+  // if landscape and height is smaller that iPad height
+  if (D_WIDTH > D_HEIGHT && D_HEIGHT < PAD_WIDTH) {
+    return false;
+  }
+
+  return true;
+})();
+
+const statusBarHeight = isLandscape => {
+  if (isIPhoneX) {
+    return isLandscape ? 0 : 44;
+  }
+
+  if (isIPad) {
+    return 20;
+  }
+
+  return isLandscape ? 0 : 20;
+};
+
+class SafeView extends Component {
+  state = {
+    touchesTop: true,
+    touchesBottom: true,
+    touchesLeft: true,
+    touchesRight: true,
+    orientation: null,
+  };
+
+  componentDidMount() {
+    InteractionManager.runAfterInteractions(() => {
+      this._onLayout();
+    });
+  }
+
+  componentWillReceiveProps() {
+    this._onLayout();
+  }
+
+  render() {
+    const { forceInset = false, isLandscape, children, style } = this.props;
+
+    if (!Platform.OS === 'ios') {
+      return <View style={style}>{this.props.children}</View>;
+    }
+
+    if (!forceInset && minor >= 50) {
+      return <SafeAreaView style={style}>{this.props.children}</SafeAreaView>;
+    }
+
+    const safeAreaStyle = this._getSafeAreaStyle();
+
+    return (
+      <View
+        ref={c => (this.view = c)}
+        onLayout={this._onLayout}
+        style={[style, safeAreaStyle]}
+      >
+        {this.props.children}
+      </View>
+    );
+  }
+
+  _onLayout = () => {
+    if (!this.view) return;
+
+    const { isLandscape } = this.props;
+    const { orientation } = this.state;
+    const newOrientation = isLandscape ? 'landscape' : 'portrait';
+    if (orientation && orientation === newOrientation) {
+      return;
+    }
+
+    const WIDTH = isLandscape ? X_HEIGHT : X_WIDTH;
+    const HEIGHT = isLandscape ? X_WIDTH : X_HEIGHT;
+
+    this.view.measureInWindow((winX, winY, winWidth, winHeight) => {
+      let realY = winY;
+      let realX = winX;
+
+      while (realY >= HEIGHT) {
+        realY -= HEIGHT;
+      }
+
+      while (realX >= WIDTH) {
```

Comment:
maybe a dumb question; when does it happen that the realX is greater than width? Also, I might be wrong, but couldn't the `while`s be replaced with modulo?

Reply:
Not a dumb question, and this one threw me for a loop. It's because in a TabNavigator the views get placed side by side and then an x-offset is changed to move the correct screen into the view port. TabView does it this way so it can easily enable swiping.

Reply:
Also, you're right, I'll update it to use modulo. :eyeroll: